File: packages/shared-types/src/capabilities.ts
import type { DatastoreEngine, DatastoreFamily } from './connection'
import type { QueryLanguage } from './workspace'

export const ADAPTER_CAPABILITIES = [
  'supports_sql_editor',
  'supports_document_view',
  'supports_graph_view',
  'supports_time_series_charting',
  'supports_explain_plan',
  'supports_transactions',
  'supports_visual_query_builder',
  'supports_key_browser',
  'supports_schema_browser',
  'supports_ttl_management',
  'supports_result_snapshots',
  'supports_query_cancellation',
  'supports_streaming_results',
  'supports_local_database_creation',
  'supports_structure_visualization',
  'supports_admin_operations',
  'supports_index_management',
  'supports_user_role_browser',
  'supports_permission_inspection',
  'supports_plan_visualization',
  'supports_query_profile',
  'supports_metrics_collection',
  'supports_cloud_iam',
  'supports_cost_estimation',
  'supports_import_export',
  'supports_backup_restore',
  'supports_vector_search',
] as const

export type AdapterCapability = (typeof ADAPTER_CAPABILITIES)[number]

export interface LocalDatabaseManifest {
  defaultExtension: string
  extensions: string[]
  canCreateEmpty: boolean
  canCreateStarter: boolean
}

export interface AdapterManifest {
  id: string
  engine: DatastoreEngine
  family: DatastoreFamily
  label: string
  maturity: 'mvp' | 'beta' | 'planned'
  capabilities: AdapterCapability[]
  defaultLanguage: QueryLanguage
  localDatabase?: LocalDatabaseManifest
}

export interface ExecutionCapabilities {
  canCancel: boolean
  canExplain: boolean
  supportsLiveMetadata: boolean
  editorLanguage: string
  defaultRowLimit: number
}

File: apps/desktop/src-tauri/src/adapters/common/capabilities.rs
use crate::domain::models::{AdapterManifest, ExecutionCapabilities};

pub(crate) const DEFAULT_PAGE_SIZE: u32 = 500;
pub(crate) const MAX_PAGE_SIZE: u32 = 5_000;

pub(crate) fn manifest(
    id: &str,
    engine: &str,
    family: &str,
    label: &str,
    default_language: &str,
    capabilities: &[&str],
) -> AdapterManifest {
    manifest_with_maturity(
        id,
        engine,
        family,
        label,
        "mvp",
        default_language,
        capabilities,
    )
}

pub(crate) fn manifest_with_maturity(
    id: &str,
    engine: &str,
    family: &str,
    label: &str,
    maturity: &str,
    default_language: &str,
    capabilities: &[&str],
) -> AdapterManifest {
    AdapterManifest {
        id: id.into(),
        engine: engine.into(),
        family: family.into(),
        label: label.into(),
        maturity: maturity.into(),
        capabilities: capabilities
            .iter()
            .map(|item| (*item).to_string())
            .collect(),
        default_language: default_language.into(),
        local_database: None,
    }
}

mod capability_sets;

pub(crate) use capability_sets::*;

pub(crate) fn sql_capabilities(can_cancel: bool, can_explain: bool) -> ExecutionCapabilities {
    ExecutionCapabilities {
        can_cancel,
        can_explain,
        supports_live_metadata: true,
        editor_language: "sql".into(),
        default_row_limit: DEFAULT_PAGE_SIZE,
    }
}
